### ISSUE TITLE
Can we do anything on minishift to help folks use HTTPS based Routes from browsers?

### ISSUE THREAD

**Issue by jstrachan** [minishift/minishift#1031]
one of the challenges of running stuff locally on your laptop is that its local and on your laptop. When running stuff on, say, openshift.io we have certs created on all public URLs and associated with the `Route` resources so that browsers can just open URLs via https fine.

When running anything on minishift over https you get that chrome dance where chrome first says its insecure - you then have to click `ADVANCED`, then click the new link that appears - which many developers get confused by. 

The same thing happens when you open the OpenShift console on minishift:
![privacy_error](https://user-images.githubusercontent.com/30140/27186275-b9ffbb70-51df-11e7-9ec2-25633bfc52d7.png)

Then after clicking `ADVANCED` you get this link at the bottom:
![privacy_error](https://user-images.githubusercontent.com/30140/27186318-db5133c6-51df-11e7-9b9f-b0c4945a3a9d.png)

Now when we start deploying apps using KeyCloak on minishift we have to use HTTPS; then we have apps that login via KeyCloak; then add in some REST APIs and you can soon have 3 or 4 URLs you have to do this same dance to get the certs accepted by the browser.

e.g. when [running fabric8 on minishift](https://github.com/fabric8io/fabric8-platform#v-4x-pre-release-development) - which is just about working now finally, we'll polish it some more then try submit as a minishift addon. There are actually about 4 separate URLs and Routes which are all exposed as HTTPS which the user has to navigate to and do this ADVANCED/click dance on.

I wonder what can we do to make it easier for folks running stuff locally on minishift to avoid this browser/cert pain?

We've done some experiments with [certbot](https://github.com/certbot/certbot)  (used to be called letsencrypt) to generate signed certs that we could then automatically update the Routes with. 
But domains need to be validated by a 'challenge' by a CA server, their CA server cant see a local minishift so that fails.

We tried running their CA server Boulder: https://github.com/letsencrypt/boulder so far that only works on minikube - I wonder if we could get it working on minishift?

But the challenge doesn't work with wildcard domains so the https challange failed, and the DNS chanllenge couldnt work either because you need to add an entry to a dns server which we can't do on kubernetes

All the apps are in a branch here https://github.com/rawlingsj/fabric8-platform/tree/certbot/apps
which if anyone wants to run boulder or certbot they can with them

So TL;DR I wonder if there's anything we can do to add some kinda certs to Routes on HTTPS on minishift to ease the developer UX of working with them?

**Comment by LalatenduMohanty** [comment]
Related #391.

**Comment by hferentschik** [comment]
> We tried running their CA server Boulder: https://github.com/letsencrypt/boulder so far that only works on minikube - I wonder if we could get it working on minishift?

Can you describe this approach a bit more?

**Comment by rawlingsj** [comment]
> Can you describe this approach a bit more?

Sure, the main problem of generating signed certs on minishift as I see it is a CA server needs to validate the requester does own the domain they're requesting a cert for.  This validation is in the form of a challenge and can happen via HTTP or DNS but the domain needs to be public for the CA server to challenge it.  I.e. it wont work when you're on a private network like when minishift runs on your laptop.

One thought was to run Boulder which is certbots CA server inside a pod so that it can see the Route `foo.nip.io` domains.  I ran into a cpl of filesystem issues running Boulder on minishift but it worked on minikube so I'm sure we can figure this out.  

Having said that boulder / certbot don't support wildcard domains so validating via HTTP doesn't work.  I.e. I have a domain `keycloak-fabric8.192.168.64.151.nip.io` that I request a signed cert for.  Boulder requires a proxy or something to server a token at a subpath of that domain i.e. http://keycloak-fabric8.192.168.64.151.nip.io/.well-known/acme-challenge.  When instead on openshift we would create a new domain for certbot to server the challenge at, i.e.  http://certbot-fabric8.192.168.64.151.nip.io/.well-known/acme-challenge which means Boulder can't validate it.  

The other option is that Boulder can use DNS to validate the domain, but in order for that to work a challenge token entry needs to be added to the DNS server.  So this doesn't work with kubernetes services as we cant add this record in.

There's more details https://certbot.eff.org/docs/using.html#id10 and it was starting to get over my head so someone with a better knowledge of this area may have better luck.  I have all the yaml to run Boulder and a certbot client https://github.com/rawlingsj/fabric8-platform/tree/certbot/apps if anyone want's to have a go.

If we can generate a signed certificate then we can write a little controller to request signed certs for any Route / Ingress rule and automatically update the Route / Ingress secret to include the keys + certs.

Is this the kinda info you're after?

**Comment by jstrachan** [comment]
I wonder if NodePorts would be easier to support on minishift? The downside would be Routes are the default openshift approach & the urls are not as clean

**Comment by jorgemoralespou** [comment]
I would think this is a topic that openshift (oc cluster and the likes) could/should solve.

cc/ @pweil- @mfojtik @smarterclayton Any idea who could help here?

**Comment by hferentschik** [comment]
> I wonder if NodePorts would be easier to support on minishift?

Sure that will work, but it won't be OpenShift idiomatic. 

We also should remember, if it would be so easy to by-pass this browser certificate checks, something would be seriously wrong here. Sure it is a pita in our particular case, but it is also what one would expect when thinking about what happens here.

AFAIK, all browsers have a way to add self signed certificates. We could document how to do that. I guess in this case one needs to do that each time a new instance is created. If we could cache/re-use the same certificate, a user would only have to do it once. Maybe that's a possibility.

**Comment by mfojtik** [comment]
@cswong for `oc cluster up`

**Comment by mfojtik** [comment]
can't we use lets encrypt here?

**Comment by jorgemoralespou** [comment]
No. As it's explained in the thread, you need a public IP.

Fixing @csrwng name.

**Comment by jstrachan** [comment]
@hferentschik agreed, a solution with Routes & https would be preferable. I'm not asking to by-pass the browser cert checks; just to improve the UX which is right now terrible for developers (accepting 5 separate pages of insecure warnings, one after the other before you can try use the web console).

If we could create a single cert and associate it with the 4 Routes in fabric8 we need (plus ideally the openshift console URL too); that might simplify things of only needing to approve a single cert?

**Comment by rawlingsj** [comment]
Agree with all that. I wonder if there's two ways forward here?
1. Short term generate a single (wildcard?) self signed cert that reduces the number of accepting browser tasks regardless of how many routes are created 
2. Longer term work with openshift / kubernetes to come up with a nice automated approach to generating signed certs for both public and private networks + automatically update openshift routes and kubernetes ingress secrets. 

Is this possible?  FWIW we're kinda in need of this pretty soon with the latest work for fabric8 on minishift almost ready. This would not only improve the getting started experience but also provide an option for openshift users to automatically generate and renew signed certs, a pretty handy feature I'm sure. If there's any ideas I can try or anything I can do to help with this please let me know.

**Comment by hferentschik** [comment]
>  agreed, a solution with Routes & https would be preferable. I'm not asking to by-pass the browser cert checks; just to improve the UX which is right now terrible for developers (accepting 5 separate pages of insecure warnings, one after the other before you can try use the web console).

Sure, I hear you. We want the same thing. 

> If we could create a single cert and associate it with the 4 Routes in fabric8 we need (plus ideally the openshift console URL too); that might simplify things of only needing to approve a single cert?

Right, that's for sure a start. 

> Longer term work with openshift / kubernetes to come up with a nice automated approach to generating signed certs for both public and private networks + automatically update openshift routes and kubernetes ingress secrets.

This is where I am atm lost. I am a bit of a novice here, but my understanding is that the problem is not signing the certs. They are always signed. The problem is that they are self signed and hence the browser cannot verify the certificate chain up to a trusted root. 

Given the dynamic nature on how the certificates are generated and the dynamic nature of the routes, I don't understand how this could be addressed in a way which makes things work out of the box. 


Is the only true way of fixing this not to let the user pass a trusted certificate to the various stages where we need a certificate and make sure this certificate is used? I guess that's where Let's Encrypt comes into play - https://letsencrypt.org. I don't think so that Let's Encrypt is going to work for host networks as we are dealing with in the Minishift case.

**Comment by rawlingsj** [comment]
>This is where I am atm lost. I am a bit of a novice here, but my understanding is that the problem is not signing the certs. They are always signed. The problem is that they are self signed and hence the browser cannot verify the certificate chain up to a trusted root.

Yes I think that is a better explanation.  We're looking to avoid self signed certs so that browsers don't need a manual approve of each https route created + avoid code needing to trust insecure certs, here's an example https://github.com/fabric8io/fabric8-init-tenant/commit/5164ca51ca23e6a169ade70db4697ef15d5703c2

> Given the dynamic nature on how the certificates are generated and the dynamic nature of the routes, I don't understand how this could be addressed in a way which makes things work out of the box.

We already have a controller that watches services and dynamically creates Ingress rules, OpenShift Routes, Kubernetes loadbalancers or switches services to NodePorts so access applications when they're deployed on different flavour clusters.  One idea is to extend this or create another more specific controller https://github.com/fabric8io/exposecontroller

> Is the only true way of fixing this not to let the user pass a trusted certificate to the various stages where we need a certificate and make sure this certificate is used? I guess that's where Let's Encrypt comes into play - https://letsencrypt.org. I don't think so that Let's Encrypt is going to work for host networks as we are dealing with in the Minishift case.

This was the basis of the investigation above (BTW letsencrypt is renamed certbot), I think there is a way, I'm just not sure how yet :)  i.e. kube-lego looks interesting https://github.com/jetstack/kube-lego/issues/26

**Comment by jstrachan** [comment]
Just saw this blog on certbot & openshift
https://blog.openshift.com/create-https-based-encrypted-urls-using-routes/

**Comment by rawlingsj** [comment]
Yeah saw that too although it's for public domains so we'd still have the same problem with nip.io when running locally on minishift.

**Comment by jstrachan** [comment]
agreed - just thought it would add a bit more background to the discussion. I do think local / nip.io based stuff is a special case - in particularly the use of nip.io probably makes it even more challenging

**Comment by hferentschik** [comment]
@jstrachan, thanks for the link. Clarifies a few things in my mind. Having a public domain is a game changer. I don't think any "serious"  Certificate Authority would and can verify a URL which is host visible only. 

Also for them it would not even be unique. There can be many users out there using/getting the same VM IP.  I think the only viable solution is to get the browser to trust your certificate. For this to work, we need a way to re-use the same certificate for the various steps where they are required. I can then configure my browser once to trust this certificate. The problem right now is that new certificates are generated all the time and I would need to configure my browser each time I create a new Minishift instance. That would not offer any improvement to the current situation where you basically just have to accept the exception.

**Comment by jstrachan** [comment]
@hferentschik generating one new certificate for each time you recreate MiniShift and then reuse it on all the routes (and for the OpenShift console) would at least mean you've only 1 certificate to approve to access any of the consoles or REST APIs running locally on minishift - which would be a huge UX improvement ;)

I guess a next level of improvement could be to allow reusing the certificate across `minishift delete; minishift start` commands?

**Comment by hferentschik** [comment]
> Right now with fabric8 on minishift we have 5 URLs to approve separately!

:-( That's bad indeed. 

> I guess a next level of improvement could be to allow reusing the certificate across minishift delete; minishift start commands?

Right, that's my thinking. We need to dig a bit into this to see what's need to make this happen. I guess this will also affect 'cluster up'. I would think we need a way to pass "our" certificate to cluster up. @csrwng , does this make sense?

**Comment by csrwng** [comment]
@hferentschik you'd basically need to update the router's default certificate with whatever certificate you want to use.

However, by default all certificates are generated using the same certificate authority (/var/lib/origin/openshift.local.config/master/ca.crt). On my Mac, if I open that file with Keychain Access and trust it, then I'm no longer asked to accept a self-signed cert on the console, metrics or any edge routes I create.

**Comment by jorgemoralespou** [comment]
@csrwng Is every instance created by "oc cluster up" have a different /var/lib/origin/openshift.local.config/master/ca.crt?

If I do what you suggest, I'll have to add that file for every instance I create. 

Would this be an option?
- Create my own "untrusted ca" (https://datacenteroverlords.com/2012/03/01/creating-your-own-ssl-certificate-authority/)
- Add that file my-ca.crt to the mac KeyChain
- Start every instance with  "oc cluster up --certificate-authority=my-ca.crt"

This would make every cluster I create trusted for me. Not?

Then, if this works, we would need a way to pass --certificate-authority to minishift to be able to do the same.

**Comment by csrwng** [comment]
@jorgemoralespou yes it will have a different one for every different config directory. If you reuse your config directory of course you'll get the same one as before.

As for what you suggest, it'd be good to try it out. Even without a flag to cluster up, it is possible to do something similar by generating your certs with `oc adm ca create-signer-cert`, save it to your favorite directory, trust it, then before running `oc cluster up --host-config-dir=[blah]` populate the host config directory with your generated signer cert.

**Comment by hferentschik** [comment]
I think we are onto something here :-). I think the way forward is the --certificate-authority option. Generate the cert once and trust it and then provide it to the bootstrap process. Seems to me the cleanest approach. Given the Minishift VMs are deleted often, we will otherwise still keep getting new master certificates. Also the --host-config-dir seems a bit of a "hack".

**Comment by csrwng** [comment]
@hferentschik I wasn't suggesting the host config dir bit as a final solution, just something to try to see if this is feasible. (Also, not much of a hack if you consider that minishift always keeps the config in the same directory inside the vm) As far as the flag goes, keep in mind that it may not be as simple as adding a single flag to cluster up. Afaik you at least need the cert, key and serial text file.

**Comment by hferentschik** [comment]
> I wasn't suggesting the host config dir bit as a final solution, just something to try to see if this is feasible

sure, git that

> As far as the flag goes, keep in mind that it may not be as simple as adding a single flag to cluster up. Afaik you at least need the cert, key and serial text file.

Let's take a look at that.

**Comment by jstrachan** [comment]
I love the idea of a persistent shared cert we can use on routes etc.

Another idea we had this morning on the walk to the office is to use the same domain name for all the various Routes but use sub-paths instead.

e.g. instead of: `https://foo-namespace.ip.nic.io/` we go with `https://namespace.ip.nic.io/foo/` so that there's only 1 URL to accept

**Comment by hferentschik** [comment]
> Another idea we had this morning on the walk to the office is to use the same domain name for all the various Routes but use sub-paths instead.

+1, I was always wondering why you creates these top level domains

**Comment by jorgemoralespou** [comment]
-1
What would be the benefit of you can trust the CA, you can create many certs as required. On the downside, I would never be able to hit those URLs without a link or deep knowledge of the application. I would trade simplicity of use for complexity on the development/admin side.

**Comment by hferentschik** [comment]
@jorgemoralespou, trying to understand you here. You are saying it helps you understand the application better if you have multiple top level domain over a single top level domain with various subpaths?

**Comment by jorgemoralespou** [comment]
@hferentschik yes. And eventually that's what most users/customers do. They provide top level domains for the applications themselves. That's what a wildcard dns provides as a benefit. You know in an easier way where to locate your app.

Think about google.

Would you rather have:

www.google.com/mail
www.google.com/docs
www.google.com/drive

or 

mail.google.com
docs.google.com
drive.google.com

Not saying that subpaths are not helpful, yes, in many situations, but it's not the default behaviour users will take. In any case, it's really subjective, and just an opinion.

**Comment by jstrachan** [comment]
@jorgemoralespou I agree, I'd prefer to use host names rather than sub paths; was just thinking of the sub paths as a workaround while we work out the cert sharing thingy

**Comment by hferentschik** [comment]
Fair enough :-)

**Comment by jorgemoralespou** [comment]
@hferentschik Hey, I took a stab at this in my "oc-cluster" wrapper since this is a function I wanted for long time. Here's what I did, but changes can be see in this commit: (https://github.com/openshift-evangelists/oc-cluster-wrapper/commit/a470fa76608635161385d73fd8bdc099d0959274)

- Created the signer

````
oc adm ca create-signer-cert \
                 --cert "$HOME/.oc/certs/ca.cert" \
                 --key "$HOME/.oc/certs/ca.key" \
                 --serial "$HOME/.oc/certs/ca.serial.txt" \
                 --name="selfsigners@openshift.local" \
                 --overwrite
````

- Added the ca to the trusted certs in my mac

````
sudo security add-trusted-cert -d -r trustRoot -k "/Library/Keychains/System.keychain" $HOME/.oc/certs/ca.cert
````

- Created the cluster with a "pre-populated" master config dir with those files.

````
cp -f $HOME/.oc/certs/ca.crt $HOME/.oc/profiles/<PROFILE>/config/master/ca.crt
cp -f $HOME/.oc/certs/ca.key $HOME/.oc/profiles/<PROFILE>/config/master/ca.key
cp -f $HOME/.oc/certs/ca.serial.txt $HOME/.oc/profiles/<PROFILE>/config/master/ca.serial.txt
oc cluster up --host-config-dir /Users/jmorales/.oc/profiles/<PROFILE>/config
````

Then, when I access the console, I no longer get a warning to trust the certificate. 

The experience is not optimal, as I would rather have flags to "oc cluster" to pass the certs, but it works.

**Comment by bpasson** [comment]
I don't know if this can be of any help to the issue, but letsencrypt is rolling out wildcard support. See https://letsencrypt.org/2017/07/06/wildcard-certificates-coming-jan-2018.html

**Comment by gbraad** [comment]
@bpasson Domain Validation needs to happen with an Internet routable endpoint. Unfortunately, and due to security constraints, we do not expose any `xip.io` or `nip.io` to be Internet.

**Comment by bpasson** [comment]
@gbraat that is true but would it work if we ship a certificate for *.nip.io and *.xip.io? It should be possible to have that validated on an external host. 

MiniShift can then just use that one on the local install, or am I wrong?

**Comment by gbraad** [comment]
@bpasson that means we share a signed certificate which is not even for a domain we own. this sounds like a big security risk...

**Comment by bpasson** [comment]
@gbraat fair point.

**Comment by bpasson** [comment]
@gbraat maybe something like plex did is an option. Have a look at: https://blog.filippo.io/how-plex-is-doing-https-for-all-its-users/

**Comment by gbraat** [comment]
guys - wrong gerard braa*t*. Please take me off this list.

Met vriendelijke groeten | Kind Regards

Gerard Braat

**Comment by gbraad** [comment]
@bpasson interesting read. they create a similar approach by having a catch-all domain, and then providing a wildcard certificate.

**Comment by bpasson** [comment]
@gbraad I thought so too. Might be an area to explore. The only concern I have is if DNS rebinding attacks are an issue for us as is described in the blog post. Since we actually never access minishift from a public website.

**Comment by cben** [comment]
Could anyone explain how certs (the untrusted ones you need to click through) are currently generated in minishift?  I don't see any /etc/origin/master/ca.crt like in a full openshift-ansible deployment.
- where is openshift API/console cert generated and stored?
  - is there a way to retrieve that cert, or whatever issued it, to trust in `curl` etc?
- is there an internal service-signer CA?

----

I think a locally persisted CA is a quite good way to go.
You only need to trust one self-signed CA in the browser, for any number of domains.

openshift-ansible generates a wildcard cert from the internal CA since https://github.com/openshift/openshift-ansible/pull/3821 (3.6, also backported 3.5).
Note that wildcard just avoids needs to generate certs at all for common case of routes, but if needed the CA can issue additional certs for *any* domain, without having to add them to browser.

Security considerations:

- Should be careful with issuing.  Imagine that for convenience, we'd somehow arrange auto-issuing certs for any custom-domain route.  Then a user carelessly instantiating a random template from the internet, could get a malicious app declaring a route for say "gmail.com", and it might steal personal data from browser...  Could say "obviously one shouldn't do that", but minishift running containers in disposable VM does *sound* like a sandbox where it's plausible to try random stuff from the internet — the implications of trusting outside the VM a CA that can issue inside the VM are far from "obvious".  Better if possible to keep the CA outside the VM, and not issue automatically beyond the initial openshift console cert + apps wildcard cert.]

- Similarly, it's important to create separate random CA for each minishift install, and not expose it to other users.  For example having one hard-coded CA for all minishifts would be very bad — once you trust the CA cert, anybody else could impersonate *any* site to you.
  For same reason, it's better practice to recommend adding the cert to your browser, not to the machine's trust store (once added it globally, minishift user can attack other users on the machine).

- If OTOH we know we only care about .nip.io addresses, I wonder if it's possible to make the CA root valid for *only* .nip.io and below to reduce risks?  (well better .nip.io *and* .xip.io to allow fallback when one is down as nip recently was)

----

Plex: well done!  I'm a bit surprised Digicert let them issue public certs for private IPs so easily, though I'm not sure if that's bad or good :)
In any case, their domain structure may be a great idea over nip/xip.io structure!
It's somewhat messy with 10.2.3.4.nip.io that anybody with same internal addresses legitimately uses exactly same domain(s).
Plex use e.g. `https://10-2-3-4.625d406a00ac415b978ddb368c0d1289.plex.direct`, where the "hash" part is unique per instance!  So no same-origin problems vs other instances, and handing the instance a wildcard cert for `*.625d406a00ac415b978ddb368c0d1289.plex.direct` is not a risk.
If anybody is working on "our own" xip clone, should consider something like this.

**Comment by cben** [comment]
-[x] Hmm, there already is a wildcard cert.
minishift 1.9.0, openshift 3.7.0, I'm seeing the api master on :8443 is serving this cert chain:
```
 0 s:/CN=10.0.2.15
   i:/CN=openshift-signer@1513073422
            X509v3 Subject Alternative Name: 
                DNS:kubernetes, DNS:kubernetes.default, DNS:kubernetes.default.svc, DNS:kubernetes.default.svc.cluster.local, DNS:localhost, DNS:openshift, DNS:openshift.default, DNS:openshift.default.svc, DNS:openshift.default.svc.cluster.local, DNS:10.0.2.15, DNS:127.0.0.1, DNS:172.17.0.1, DNS:172.30.0.1, DNS:192.168.99.100, IP Address:10.0.2.15, IP Address:127.0.0.1, IP Address:172.17.0.1, IP Address:172.30.0.1, IP Address:192.168.99.100
...
 1 s:/CN=openshift-signer@1513073422
   i:/CN=openshift-signer@1513073422
```
and the router on :433 is serving:
```
 0 s:/CN=*.192.168.99.100.nip.io
   i:/CN=openshift-signer@1513073422
...
 1 s:/CN=openshift-signer@1513073422
   i:/CN=openshift-signer@1513073422
...
 2 s:/CN=openshift-signer@1513073422
   i:/CN=openshift-signer@1513073422
```

And after some searching I found the openshift-signer CA cert to trust, it's inside the origin container:
```
minishift ssh docker exec origin /bin/cat ./openshift.local.config/master/ca.crt
```

- [ ] So seems CA & cert is generated inside origin container, so not persisted between minishift VMs.

(also found a few certs in ~/.minishift but they were all different, Subject: O = bpaskinc, O = "bpaskinc.<bootstrap>", O = bpaskinc.minishift, where bpaskinc is my username.)

- [x] I was able to `oc login --certificate-authority=above.crt https://192.168.99.100:8443`.
- will check after lunch and edit this if trusting CA in browser is enough for apps

**Comment by jorgemoralespou** [comment]
nip.io could be used with letsencrypt.

https://github.com/resmo/nip.io/issues/2

**Comment by jorgemoralespou** [comment]
keep

**Comment by yadavsachin0995** [comment]
Thank you cben.
